### ISSUE TITLE
Automatically narrow `__package__` to `str` for package modules

### ISSUE THREAD

**Issue by effigies** [microsoft/pyright#9565]
**Is your feature request related to a problem? Please describe.**

This is a follow-up to https://github.com/microsoft/pyright/issues/7408, and may just be a clarification.

I believe that whether a file is a package, and thus whether `__package__` is a `str`, can be inferred in some cases. For example, in `src/mypkg/data/__init__.py`, I might have:

```python
from importlib.resources import files

data_files = files(__package__)
```

which allows me to do something like:

```python
from mypkg.data import data_files
```

Following #7409, this now requires `assert __package__` first. A library that tries to be conscientious about import times cannot control whether end-users use `-O` to remove runtime assertions.

**Describe the solution you’d like**

I would like to automatically narrow the type of `__package__` to `str` in cases where it can be known. I don't know if this is plausible in pyright's design.

**Comment by erictraut** [comment]
I don't think this can be determined statically. The type of `__package__` depends on how and where a Python file is executed. There are some heuristics that could be used, but these would be incorrect some of the time.

The type of `__package__` comes from the `types.ModuleType` type definition in typeshed. Overriding type definitions provided by typeshed is not something that I like to do. It tends to lead to maintenance and compatibility issues over time. 

For these reasons, I'm going to reject this enhancement request at this time. I'm open to revisiting this decision in the future if there's sufficient signal from other pyright users that this feature is important to them.

In the meantime, you can use an `assert __package__ is not None` in your code to document and validate your assumption — both for human readers and for type checkers.

**Comment by effigies** [comment]
Just to follow up, for anybody who might run across this issue:

* `__package__` has been deprecated and shouldn't be relied upon anyway. The equivalent `__spec__.parent` is documented as possibly being `None`.
* `__spec__.name` is guaranteed to be a string, and for `__init__.py` files, it is the same as `__spec__.parent`.

It would seem reasonable to me for pyright to narrow `__spec__.parent` iff it can be statically determined to be a `str`, but I have never found a use for `__package__` outside `__init__.py` files, so the existence of `__spec__.name` makes this moot, as far as I'm concerned.
